src/brush.jsx

```diff
@@ -9,75 +9,371 @@
+            fill: "white",
+            opacity: 0,
+            cursor
+        };
+        return (
+            <rect
+                ref="overlay"
+                x={0} y={0}
+                width={width} height={height}
+                style={overlayStyle}
+                onMouseDown={this.handleOverlayMouseDown}
+                onMouseUp={this.handleMouseUp}
+                onClick={this.handleClick}
+                clipPath={this.props.clipPathURL} />
+        );
     },
 
-    componentWillReceiveProps(nextProps) {
-        const timeScale = nextProps.timeScale;
-        const timeRange = nextProps.timeRange;
-        if (scaleAsString(this.props.timeScale) !== scaleAsString(timeScale) ||
-            this.props.timeRange !== timeRange) {
-            this.renderBrush(timeScale, timeRange);
+    renderBrush() {
+        const { timeRange, timeScale, height, style } = this.props;
+
+        if (!timeRange) {
+            return (
+                <g />
+            );
+        }
+
+        let cursor;
+        switch (this.state.brushingInitializationSite) {
+            case "handle-right":
+            case "handle-left":
+                cursor = "ew-resize";
+                break;
+            case "overlay":
+                cursor = "crosshair";
+                break;
+            default:
+                cursor = "move";
+        }
+
+        // Style of the brush area
+        const brushDefaultStyle = {
+            fill: "#777",
+            fillOpacity: 0.3,
+            stroke: "#fff",
+            shapeRendering: "crispEdges",
+            cursor
+        };
+        const brushStyle = merge(true, brushDefaultStyle, style);
+
+        if (!this.viewport().disjoint(timeRange)) {
+            const range = timeRange.intersection(this.viewport());
+            const begin = range.begin();
+            const end = range.end();
+            const [ x, y ] = [timeScale(begin), 0];
+            const endPos = timeScale(end);
+            let width = endPos - x;
+            if (width < 1) {
+                width = 1;
+            }
+
+            const bounds = {x, y, width, height};
+
+            return (
+                <rect
+                    {...bounds}
+                    style={brushStyle}
+                    pointerEvents="all"
+                    onMouseDown={this.handleBrushMouseDown}
+                    onMouseUp={this.handleMouseUp}
+                    clipPath={this.props.clipPathURL} />
+            );
         }
     },
 
-    shouldComponentUpdate() {
-        return false;
+    renderHandles() {
+        const { timeRange, timeScale, height } = this.props;
+
+        if (!timeRange) {
+            return (
+                <g />
+            );
+        }
+
+        // Style of the handles
+        const handleStyle = {
+            fill: "white",
+            opacity: 0,
+            cursor: "ew-resize"
+        };
+
+        if (!this.viewport().disjoint(timeRange)) {
```

Comment:
If this is disjoint it returns undefined.  That shouldn't happen, but would it be worthwhile to return an empty <g /> and maybe put a warning in the console?